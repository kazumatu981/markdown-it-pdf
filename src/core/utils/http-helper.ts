import http from 'http';
import { Logger } from '../../common/logger';
import type { Range, ServerPortOptions } from '../../common/configure';

//#region constants
const privatePortRange = {
    min: 49152,
    max: 65535,
};
//#endregion

//#region types and interfaces
export interface ServerPort {
    /**
     * The port that the server is listening on.
     */
    port: number;
    /**
     * The http server that is listening on the specified port.
     */
    httpServer: http.Server;
}
//#endregion

//#region functions

//#region exported functions
/**
 * Tries to listen on a given port. If the port is not specified, it tries to
 * listen on a random port within a specified range. If the specified port is
 * in use, it retries the specified number of times.
 *
 * @param {number | undefined} port The port to listen on. If not specified, it will try to listen on
 * a random port within the specified range.
 * @param {ServerPortOptions | undefined} options The options for listening on a port.
 * @param {Logger | undefined} logger The logger to log the events.
 * @returns {Promise<ServerPort | undefined>} A promise that resolves to the port that the server is listening on
 * or undefined if the specified port is in use and cannot be listened on.
 */
export async function tryToListen(
    port?: number,
    options?: ServerPortOptions,
    logger?: Logger
): Promise<ServerPort | undefined> {
    let serverPort = undefined;
    if (port != undefined) {
        logger?.debug('Try to Listen on port %d', port);
        serverPort = await tryToListenCore(port, logger);
    } else {
        logger?.debug('Try to Listen on port random mode');
        for (let i = 0; i < (options?.retry ?? 10); i++) {
            port = getRandom(options?.range);
            serverPort = await tryToListenCore(port, logger);
            if (serverPort !== undefined) {
                break; // success
            }
        }
    }
    return serverPort;
}
//#endregion

//#region internal functions
/**
 * Generates a random number within a specified range.
 *
 * @param {Range | undefined} range - The range within which the random number
 * should be generated. If not specified, the function will use the default range
 * of 49152 to 65535.
 * @returns {number} A random number within the specified range.
 * @throws {Error} If the range is invalid (min > max).
 */
function getRandom(range?: Range): number {
    // Set the default range if not specified
    const candidate = {
        min: range?.min ?? privatePortRange.min,
        max: range?.max ?? privatePortRange.max,
    };

    // Check if the range is valid
    if (candidate.min > candidate.max) {
        throw new Error('Invalid range');
    }

    // Generate a random number within the range
    return Math.floor(
        Math.random() * (candidate.max - candidate.min) + candidate.min
    );
}

/**
 * Tries to listen on a given port. If the port is in use, it rejects the
 * promise.
 *
 * @param {number} port The port to listen on.
 * @param {Logger | undefined} logger The logger to log the events.
 * @returns {Promise<ServerPort | undefined>} A promise that resolves to the port that the server is listening on
 * or undefined if the specified port is in use and cannot be listened on.
 */
function tryToListenCore(
    port: number,
    logger?: Logger
): Promise<ServerPort | undefined> {
    // Create a new http server
    const server = http.createServer();

    // Create a promise that resolves to the port that the server is listening on
    // or undefined if the specified port is in use and cannot be listened on
    return new Promise((resolve) => {
        // Start listening on the specified port
        server.listen(port, () => {
            // Log the reserved port
            logger?.debug('reserved port: %d', port);
            // Resolve the promise with the port and the server
            resolve({
                port,
                httpServer: server,
            });
        });
        // If there is an error while listening on the port, log the error and
        // resolve the promise with undefined
        server.on('error', (error) => {
            logger?.debug('listen error: %s', error.message);
            logger?.trace(error.stack);
            logger?.warn('port %d may be used.', port);
            resolve(undefined);
        });
    });
}
//#endregion

//#endregion
