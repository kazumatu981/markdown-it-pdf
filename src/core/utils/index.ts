export * from './http-helper';
export * from './path-resolver';
