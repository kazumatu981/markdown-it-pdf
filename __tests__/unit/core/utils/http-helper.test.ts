import { jest, expect, describe, it, afterEach } from '@jest/globals';
import { mockLogger } from '../../../utils/mock-logger';

import { tryToListen } from '../../../../src/core/utils/http-helper';

describe('CoreLibrary Unit Tests - http-helper', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should return a valid port', async () => {
        const serverPort = await tryToListen();
        expect(serverPort).not.toBeUndefined();
        serverPort?.httpServer?.close();
    });

    it('should return a valid port - with logger', async () => {
        const serverPort = await tryToListen(undefined, undefined, mockLogger);
        expect(serverPort).not.toBeUndefined();
        expect(mockLogger.debug).toBeCalled();
        serverPort?.httpServer?.close();
    });

    it('should return a valid port even if the port is already in use', async () => {
        const serverPort1 = await tryToListen(3000);
        const serverPort = await tryToListen();
        expect(serverPort).not.toBeUndefined();
        serverPort?.httpServer?.close();
        serverPort1?.httpServer?.close();
    });
    it('should return undefined if the port is already in use', async () => {
        const serverPort1 = await tryToListen(3000);
        const serverPort = await tryToListen(3000);
        expect(serverPort).toBeUndefined();
        serverPort?.httpServer?.close();
        serverPort1?.httpServer?.close();
    });
    it('should return undefined if the port is already in use - with logger', async () => {
        const serverPort1 = await tryToListen(3000, undefined, mockLogger);
        const serverPort = await tryToListen(3000, undefined, mockLogger);

        expect(mockLogger.error).toMatchSnapshot(
            'on to have been duplicated port, logger error'
        );
        expect(mockLogger.warn).toMatchSnapshot(
            'on to have been duplicated port, logger warn'
        );
        expect(mockLogger.info).toMatchSnapshot(
            'on to have been duplicated port, logger info'
        );
        expect(mockLogger.debug).toMatchSnapshot(
            'on to have been duplicated port, logger debug'
        );
        expect(mockLogger.trace).toBeCalled();

        serverPort?.httpServer?.close();
        serverPort1?.httpServer?.close();
    });
    it('should throw on error, if th range is invalid', async () => {
        expect(async () => {
            await tryToListen(undefined, { range: { min: 3001, max: 3000 } });
        }).rejects.toThrow();
    });
});
