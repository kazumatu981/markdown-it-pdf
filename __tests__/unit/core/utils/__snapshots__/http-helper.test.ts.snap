// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreLibrary Unit Tests - http-helper should return undefined if the port is already in use - with logger: on to have been duplicated port, logger debug 1`] = `
[MockFunction] {
  "calls": [
    [
      "Try to Listen on port %d",
      3000,
    ],
    [
      "reserved port: %d",
      3000,
    ],
    [
      "Try to Listen on port %d",
      3000,
    ],
    [
      "listen error: %s",
      "listen EADDRINUSE: address already in use :::3000",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - http-helper should return undefined if the port is already in use - with logger: on to have been duplicated port, logger error 1`] = `[MockFunction]`;

exports[`CoreLibrary Unit Tests - http-helper should return undefined if the port is already in use - with logger: on to have been duplicated port, logger info 1`] = `[MockFunction]`;

exports[`CoreLibrary Unit Tests - http-helper should return undefined if the port is already in use - with logger: on to have been duplicated port, logger warn 1`] = `
[MockFunction] {
  "calls": [
    [
      "port %d may be used.",
      3000,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
